kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+            pNode->HasDofFor(GetComponentVariable(rVariable, 2))) {
+
+            rCloud_of_dofs_x[i] = pNode->pGetDof(GetComponentVariable(rVariable, 0));
+            rCloud_of_dofs_y[i] = pNode->pGetDof(GetComponentVariable(rVariable, 1));
+            rCloud_of_dofs_z[i] = pNode->pGetDof(GetComponentVariable(rVariable, 2));
+
+        } else {
+            KRATOS_ERROR << "The node with ID " << pNode->Id() << " does not have the required DOFs for the variable " << rVariable << std::endl;
+        }
+        noalias(row(rCloud_of_nodes_coordinates, i)) = pNode->Coordinates();
+    }
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignRotationToNodes(
+    NodesContainerType pNodes,
+    Matrix RotationMatrix
+    )
+{
+    //TODO: Do it in parallel
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(pNodes.GetContainer());
+
+    // Assign rotation to given nodes
+    for(unsigned int i = 0; i < nodes_array.size(); i++){
+        auto coordinate_vector = nodes_array[i]->GetInitialPosition().Coordinates();
+        Vector rotated_position(coordinate_vector.size());
+        noalias(rotated_position) = prod(RotationMatrix,coordinate_vector);
+        nodes_array[i]->X() = rotated_position[0];
+        nodes_array[i]->Y() = rotated_position[1];
+        nodes_array[i]->Z() = rotated_position[2];
+    }
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignMPCsToNodes(
+    NodesContainerType pNodes,
+    double const Radius,
+    ModelPart& rComputingModelPart,
+    const Variable<double>& rVariable,
+    double const MinNumOfNeighNodes
+    )
+{
+    KRATOS_TRY;
+
+    BuiltinTimer build_and_assign_mpcs;
+
+    int prev_num_mpcs = rComputingModelPart.NumberOfMasterSlaveConstraints();
+
+    NodesContainerType::ContainerType& nodes_array = const_cast<NodesContainerType::ContainerType&>(pNodes.GetContainer());
+
+    ModelPart::MasterSlaveConstraintType const& r_clone_constraint = KratosComponents<MasterSlaveConstraint>::Get("LinearMasterSlaveConstraint");
+
+    // Create a concurrent queue for constraints
+    using MasterSlaveConstarintQueue = moodycamel::ConcurrentQueue<ModelPart::MasterSlaveConstraintType::Pointer>;
+
+    MasterSlaveConstarintQueue concurrent_constraints;
+
+    // Declare a counter variable outside the loop as std::atomic<int>
+    std::atomic<int> i(0);
+
+    block_for_each(nodes_array, [&](Node<3>::Pointer& pNode) {
+
+        double local_radius = Radius;
+
+        ResultNodesContainerType r_result(0);
+
+        while (r_result.size() < MinNumOfNeighNodes) {
+            r_result.clear();
+            SearchNodesInRadiusForNode(pNode, local_radius, r_result);
+            local_radius += Radius;
+        }
+
+        // Get Dofs and Coordinates
+        DofPointerVectorType r_cloud_of_dofs, r_slave_dof;
+        Matrix r_cloud_of_nodes_coordinates;
+        array_1d<double, 3> r_slave_coordinates;
```

Comment:
You're allocating/deallocating these for each node. Declare them outside the loop.

Reply:
Master-Slave constraints require rSlaveDofs to be a DofPointerVectorType. To do r_slave_dofs thread_local I need to resize and assign to Dof<double>::Pointer. Or:
rCloudOfDofs.resize(1); 
rCloudOfDofs[0] = pNode->pGetDof(rVariable);
Which I dont like. I will fix it.